torch/nn/modules/loss.py

```diff
@@ -1246,6 +1253,144 @@ def forward(self, anchor: Tensor, positive: Tensor, negative: Tensor) -> Tensor:
                                      eps=self.eps, swap=self.swap, reduction=self.reduction)
 
 
+class TripletMarginLossWithDistance(_Loss):
+    r"""Creates a criterion that measures the triplet loss given input
+    tensors :math:`a`, :math:`p`, and :math:`n` (representing anchor,
+    positive, and negative examples, respectively); and a real-valued function
+    between them.
+
+    The unreduced loss (i.e., with `reduction` set to `'none'`)
+    can be described as:
+
+    .. math::
+        \ell(a, p, n) = L = \{l_1,\dots,l_N\}^\top, \quad
+        l_i = \max \{d(a_i, p_i) - d(a_i, n_i) + {\rm margin}, 0\}
+
+    where :math:`N` is the batch size; :math:`d` is a real-valued function quantifying
+    the separation between two tensors, referred to as `distance_function`;
+    and :math:`margin` is a non-negative margin enforced between the positive and
+    negative distances.  The input tensors have :math:`N` elements each and can be of
+    any shape that the distance function can handle.
+
+    If :attr:`reduction` is not ``'none'``
+    (default ``'mean'``), then:
+
+    .. math::
+        \ell(x, y) =
+        \begin{cases}
+            \operatorname{mean}(L), &  \text{if reduction} = \text{`mean';}\\
+            \operatorname{sum}(L),  &  \text{if reduction} = \text{`sum'.}
+        \end{cases}
+
+    See also :class:`~torch.nn.TripletMarginLoss`, which computes the triplet
+    loss for input tensors using the :math:`l_p` distance as the distance function.
+
+    Args:
+        distance_function (callable, optional): A distance function between two Tensors which,
+            if specified, will be used instead of the pairwise distance. If not specified,
+            `nn.PairwiseDistance` will be used.  Default: ``None``
+        is_similarity_function (bool, optional): Whether `distance_function` represents a
```

Comment:
I am not sure about this flag as it makes the API quite a bit harder to understand.
Do you have an example where it is required? How hard will it be for the user to just change the function it provides to just make it a similarity?

Reply:
This allows, e.g., `nn.CosineSimilarity()` to be used as a drop-in.  Otherwise, the user would presumably need to define a separate function that computes distance as 1 - similarity